### PR TITLE
Support Numba Runtime in RBC

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This PR rewrites the Numba Runtime using LLVM.

### CODE REVIEW

File: rbc/nrt.py [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+atomic_i64_t = ir.global_context.get_identified_type('atomic_i64')
+atomic_i64_t.set_body(i64)
+
+# MemInfo struct
+MemInfo_t = ir.global_context.get_identified_type('Struct.MemInfo')
+# First argument ought to be "atomic_i64_t" instead of i64
+MemInfo_t.set_body(i64, i8p, i8p, i8p, i64, i8p)
+MemInfo_ptr_t = MemInfo_t.as_pointer()
+
+sizeof_MemInfo_t = i64(48)
```
Comment (guilhermeleobas) [inline rbc/nrt.py:44]: This should be computed using target context

File: rbc/nrt.py [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+# };
+
+# atomic_i64_t is the llvm representation of std::atomic<int64>, according to godbolt.com
+atomic_i64_t = ir.global_context.get_identified_type('atomic_i64')
+atomic_i64_t.set_body(i64)
+
+# MemInfo struct
+MemInfo_t = ir.global_context.get_identified_type('Struct.MemInfo')
+# First argument ought to be "atomic_i64_t" instead of i64
+MemInfo_t.set_body(i64, i8p, i8p, i8p, i64, i8p)
```
Comment (guilhermeleobas) [inline rbc/nrt.py:41]: Replace the first `i64` by `atomic_i64_t`

Comment (guilhermeleobas) [inline rbc/nrt.py:41]: `size_t` ought to be computed using target extension and target info

File: rbc/nrt.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+            not_null = cgutils.is_not_null(builder, dtor)
+            with cgutils.if_likely(builder, not_null):
+                info_ = NULL
+                builder.call(dtor, [ptr, size, info_])
+        builder.ret_void()
+
+    def define_nrt_allocate_meminfo_and_data(self, builder, args):
+        with self.nrt_debug_ctx():
+            [size, mi_out, allocator] = args
+            # TODO: replace this by `get_abi_sizeof(...)`
```
Comment (guilhermeleobas) [inline rbc/nrt.py:414]: Address this TODO

File: rbc/nrt.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+                with then:
+                    bb_then = builder.basic_block
+                    mi_ptr = builder.bitcast(ptr, MemInfo_ptr_t, name='meminfo_ptr')
+                    data = self._get_from_meminfo(mi_ptr, 'data')
+                    self.NRT_Debug("data: %p\n", data)
+                with otherwise:
+                    # there is a specific case where ptr is NULL and doesn't
+                    # crash Numba
+                    bb_else = builder.basic_block
+                    data_null = NULL
```
Comment (guilhermeleobas) [inline rbc/nrt.py:329]: Document this odd corner case

File: rbc/tests/heavydb/test_nrt_list.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    # [X] List.count
+    # [X] List.sort
+    # [X] List.reverse
+    # [X] List.copy
+
+    @heavydb("i32(i32)", devices=['cpu'])
+    def list_append(t):
+        lst = list()
+        lst.append('abc')
+        return len(lst)
```
Comment (guilhermeleobas) [inline rbc/tests/heavydb/test_nrt_list.py:37]: Not related to this test, but introduce an `array.to_list()` and `array.to_set()` methods

File: rbc/tests/heavydb/test_nrt_string.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        elif method == 'title':
+            r = s.title()
+        else:
+            r = ''
+        return TextEncodingNone(r)
+
+    @heavydb('bool(TextEncodingNone, TextEncodingNone, TextEncodingNone)')
+    def test_endswith(method, t, value):
+        s = t.to_string()
+        w = value.to_string()
```
Comment (guilhermeleobas) [inline rbc/tests/heavydb/test_nrt_string.py:56]: Introduce a test to ensure `to_string()` method is working correctly.

File: rbc/unicodetype_db.ll [hunk 7]
```diff
@@ -0,0 +1,2297 @@
+
```
Comment (guilhermeleobas) [inline rbc/unicodetype_db.ll:1]: Document how this file was generated and how to do it in the future if needed.

File: rbc/nrt.py [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+        self.define()
+        self.set_hashsecrets()
+
+        if self.verbose:
+            print(self.module)
+
+    def set_hashsecrets(self):
+        for v in hashsecret.values():
+            gv = ir.GlobalVariable(self.module, i64, v.symbol)
+            gv.initializer = i64(v.value.value)
```
Comment (guilhermeleobas) [inline rbc/nrt.py:167]: hashsecrets should go to `unicodetype_db.ll`

File: rbc/nrt.py [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+realloc = 'realloc'
+
+nrt_debug_incr = 'nrt_debug_incr'
+nrt_debug_decr = 'nrt_debug_decr'
+nrt_debug_spaces = 'nrt_debug_spaces'
+nrt_global_var = '__nrt_global_var'
+
+function_types = {
+    NRT_MemInfo_init: (ir.FunctionType(void, [MemInfo_ptr_t, i8p, i64, i8p, i8p, i8p]),
+                       ('mi', 'data', 'size', 'dtor', 'dtor_info', 'external_allocator')),
```
Comment (guilhermeleobas) [inline rbc/nrt.py:90]: Some of the `i64` values here are actually `size_t`. Replace it by the correct definition.

**Review by guilhermeleobas — COMMENTED** [review]
Notes to myself

File: rbc/nrt.py [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+        # allocator is always null as we don't use this argument in RBC/HeavyDB
+        with self.nrt_debug_ctx():
+            [size, allocator_] = args
+
+            ptr = builder.call(self.malloc, [size])
+            # allocate_varlen_buffer calls malloc using the formula:
+            #   malloc((elem_count + 1) * elem_size)
+            # elem_count = builder.sub(size, i64(1), name='element_count')
+            # elem_size = i64(1)
+            # ptr = builder.call(self.allocate_varlen_buffer, [elem_count, elem_size])
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/nrt.py]: Changed in [[1305c4a](https://github.com/xnd-project/rbc/pull/531/commits/1305c4a57a5d0c1106572e8c78d220ef3b809f35)](https://github.com/xnd-project/rbc/pull/531/commits/1305c4a57a5d0c1106572e8c78d220ef3b809f35)

File: rbc/heavydb/text_encoding_none.py [hunk 11]
```diff
@@ -198,12 +182,58 @@ def heavydb_text_encoding_none_constructor_memcpy(context, builder, sig, args):
...
+            cgutils.raw_memcpy(builder, dst=p2, src=uni_str.data,
+                               count=uni_str.length, itemsize=i64(1))
+            # string is null terminated
+            builder.store(i8(0), builder.gep(p2, [uni_str.length]))
+            n2 = i8(0)
+
+    # {i8*, i64, i8}
+    #  ^^^
+    ptr = builder.phi(i8p)
+    ptr.add_incoming(p1, bb_then)
```
Comment (guilhermeleobas) [inline rbc/heavydb/text_encoding_none.py:218]: No need to create `p1`
```suggestion
    ptr.add_incoming(NULL, bb_then)
```

File: rbc/heavydb/array.py [hunk 12]
```diff
@@ -168,6 +174,19 @@ def heavydb_array_constructor(context, builder, sig, args):
...
+@extending.lower_builtin(Array, nb_types.List)
+def heavydb_array_ctor_list(context, builder, sig, args):
+    dtype = sig.args[0].dtype
+
+    def ctor(lst):
+        sz = len(lst)
+        arr = Array(sz, dtype)
+        for i in range(sz):
+            arr[i] = lst[i]
+        return arr
```
[the remark this answers is not in the document]
Comment (brenocfg) [reply rbc/heavydb/array.py:197]: Depending on how complicated the IR becomes, there's a chance that LLVM will optimize it on its own with `-memcpyopt`: https://llvm.org/docs/Passes.html#memcpyopt-memcpy-optimization

File: rbc/heavydb/text_encoding_none.py [hunk 13]
```diff
@@ -141,51 +152,24 @@ def heavydb_text_encoding_none_constructor(context, builder, sig, args):
...
+    module = builder.module
 
-    # create function and prevent llvm from optimizing it
-    fnty = ir.FunctionType(ir.VoidType(), [arr.type, src_data.type, sz.type])
-    func = ir.Function(module, fnty, fn_name)
-    func.attributes.add('noinline')
+    name = 'copy_bytes_fn'
+    fnty = ir.FunctionType(void, [i8p, i8p, i64])
+    fn = cgutils.get_or_insert_function(module, fnty, name)
+    fn.linkage = 'linkonce'
```
Comment (brenocfg) [inline rbc/heavydb/text_encoding_none.py:162]: Why `linkonce` here? Can we have multiple definitions of this function in a module?

File: rbc/heavydb/text_encoding_none.py [hunk 14]
```diff
@@ -198,12 +182,58 @@ def heavydb_text_encoding_none_constructor_memcpy(context, builder, sig, args):
     return fa._getpointer()
 
 
+@extending.lower_builtin(TextEncodingNone, nb_types.UnicodeType)
+def text_encoding_none_unicode_ctor(context, builder, sig, args):
+    # Note to the future:
+    # You shall be tempted to use "context.make_helper",
```
Comment (brenocfg) [inline rbc/heavydb/text_encoding_none.py:189]: You said this but used `context.make_helper` in the very first following line of code 😅. I get why `extract_value` is a bad idea, but why the other two? It's not very clear what they do on first glance.

File: rbc/heavydb/text_encoding_none.py [hunk 15]
```diff
@@ -198,12 +182,58 @@ def heavydb_text_encoding_none_constructor_memcpy(context, builder, sig, args):
...
+
+    eq = builder.icmp_signed('==', uni_str.length, uni_str.length.type(0))
+    with builder.if_else(eq, likely=False) as (then, otherwise):
+        with then:
+            bb_then = builder.basic_block
+            p1 = cgutils.get_null_value(i8p)
+            n1 = i8(1)
+        with otherwise:
+            bb_else = builder.basic_block
+            p2 = allocate_varlen_buffer(builder, uni_str.length, i64(1))
```
Comment (brenocfg) [inline rbc/heavydb/text_encoding_none.py:209]: Is `uni_str.length` here the length of the string **including** the `\0` that terminates the string? Because in C `strlen` will usually be 1 byte short since it doesn't consider that:

```
sizeof("blah") = 5
strlen("blah") = 4
```

So if NRT is written in C I'd expect it to behave similarly. If so then we'd be allocating 1 byte short here as well, no?

File: rbc/irtools.py [hunk 16]
```diff
@@ -36,7 +38,13 @@ def get_called_functions(library, funcname=None):
...
         for instruction in block.instructions:
             if instruction.opcode == 'call':
-                name = list(instruction.operands)[-1].name
+                instr = list(instruction.operands)[-1]
+                name = instr.name
+                # ignore in a call to a function pointer
+                if 'bitcast' in str(instr):
+                    continue
+                if name == '':
+                    continue
```
Comment (brenocfg) [inline rbc/irtools.py:47]: In the LLVM C++ API, this is usually done by checking if `CallInst->getCalledFunction() == nullptr || CallInst->getCalledFunction()->isDeclaration()`, does `llvmlite` not have similar methods?

File: rbc/nrt.py [hunk 17]
```diff
@@ -0,0 +1,670 @@
...
+
+        for name in dir(self):
+            if not name.startswith("define_"):
+                continue
+
+            if name in _debug_functions and not self.debug_nrt:
+                continue
+
+            meth = getattr(self, name)
+            fn_name = name[len("define_"):]
```
Comment (brenocfg) [inline rbc/nrt.py:337]: This is some clever wizardry

File: rbc/nrt.py [hunk 18]
```diff
@@ -0,0 +1,670 @@
...
+        with self.nrt_debug_ctx():
+            # Because the previously reallocated memory might be freed by
+            # realloc, RBC cannot call it in this function. Therefore, instead
+            # of reallocation, RBC allocates a new block of memory and returns
+            # it to the caller. The caller is responsible for copying memory
+            # from the old buffer into the most recent allocated one."
+            [ptr, size] = args
+            new_ptr = builder.call(self.NRT_Allocate_External, [size, NULL])
+            self.NRT_Debug("ptr=%p size=%zu -> new_ptr=%p", ptr, size, new_ptr)
+        builder.ret(new_ptr)
```
Comment (brenocfg) [inline rbc/nrt.py:414]: Is there any reason why `realloc` can't just be a wrapper around `malloc` + `memcpy` then? That would remove the weirdness about the caller being responsible for copying, which isn't the case for `realloc`. I think it's preferable to keep to the original function's API as much as possible.

File: rbc/heavydb/text_encoding_none.py (hunk 15 above)
Comment (brenocfg) [reply rbc/heavydb/text_encoding_none.py:209]: I just saw your comment later on that `allocate_varlen_buffer` allocates `len + 1` bytes, so this makes sense now :). Disregard the above comment.

**Review by brenocfg — COMMENTED** [review]
Overall looks good, this is awesome stuff @guilhermeleobas!

I have a few questions/comments, mostly related to the API of the NRT functions. Will wait for your feedback on those before giving explicit approval.

File: rbc/heavydb/text_encoding_none.py (hunk 15 above)
Comment (guilhermeleobas) [reply rbc/heavydb/text_encoding_none.py:209]: Yes! It allocates one more byte, which we use to store the null terminator

File: rbc/irtools.py (hunk 16 above)
Comment (guilhermeleobas) [reply rbc/irtools.py:47]: Unfortunately, no. The available methods for a ValueRef are:

```python
(Pdb++) print(instruction)
  %.32 = call i8* bitcast (%Struct.MemInfo.5* (i64, i8*)* @NRT_MemInfo_new_varsize_dtor to i8* (i64, i8*)*)(i64 64, i8* bitcast (void (i8*)* @.dtor.list.int64 to i8*))
(Pdb++) instruction.has_name
*** AttributeError: 'ValueRef' object has no attribute 'has_name'
(Pdb++) vars(instruction)
{'_kind': 'instruction', '_parents': {'module': <llvmlite.binding.module.ModuleRef object at 0x7feb8bf0bf10>, 'function': <llvmlite.binding.value.ValueRef object at 0x7feb8bf0bfa0>, 'block': <llvmlite.binding.value.ValueRef object at 0x7feb8b62e7a0>}, '_ptr': <llvmlite.binding.ffi.LP_LLVMValue object at 0x7feb8b832140>, '_as_parameter_': <llvmlite.binding.ffi.LP_LLVMValue object at 0x7feb8b832140>, '_capi': <llvmlite.binding.ffi._lib_wrapper object at 0x7fec59bedc00>}
(Pdb++) dir(instruction)
['__bool__', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__nonzero__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_as_parameter_', '_capi', '_closed', '_dispose', '_kind', '_owned', '_parents', '_ptr', 'add_function_attribute', 'arguments', 'attributes', 'block', 'blocks', 'close', 'closed', 'detach', 'function', 'instruction', 'instructions', 'is_argument', 'is_block', 'is_declaration', 'is_function', 'is_global', 'is_instruction', 'is_operand', 'linkage', 'module', 'name', 'opcode', 'operands', 'storage_class', 'type', 'visibility']
```

File: rbc/heavydb/text_encoding_none.py (hunk 13 above)
Comment (guilhermeleobas) [reply rbc/heavydb/text_encoding_none.py:162]: This function is not actually used anymore and we can remove it. But answering your question. Without the linkage as `linkonce`, we would get multiple definition errors.

Comment (brenocfg) [reply rbc/heavydb/text_encoding_none.py]: Weird, why so? Also if it's not used anymore then I guess removing it is ok?

File: rbc/irtools.py (hunk 16 above)
Comment (brenocfg) [reply rbc/irtools.py]: Bummer :(

File: rbc/nrt.py (hunk 18 above)
Comment (brenocfg) [reply rbc/nrt.py:420]: Ahh that makes sense.

File: rbc/heavydb/text_encoding_none.py (hunk 13 above)
Comment (guilhermeleobas) [reply rbc/heavydb/text_encoding_none.py]: Yeah, just removed it.

File: rbc/heavydb/text_encoding_none.py (hunk 14 above)
Comment (brenocfg) [reply rbc/heavydb/text_encoding_none.py:168]: Ah yeah, it's pretty safe to assume any code that handles/creates structs directly in LLVM is intended for stack-allocated struct objects and not pointers.

**Comment by guilhermeleobas** [comment]
Pinging @mflaxman10 for visibility

File: rbc/tests/heavydb/test_nrt_set.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    # [X] set.intersection
+    # [X] set.intersection_update
+    # [X] set.isdisjoint
+    # [X] set.issubset
+    # [X] set.issuperset
+    # [X] set.pop
+    # [X] set.remove
+    # [X] set.symmetric_difference
+    # [X] set.symmetric_difference_update
+    # [ ] set.union
```
Comment (pearu) [inline rbc/tests/heavydb/test_nrt_set.py:36]: What blocks supporting `set.union`?

File: rbc/unicodetype_db.h [hunk 20]
```diff
@@ -0,0 +1,46 @@
+#include <stdint.h>
+#include "path/to/numba/_unicodetype_db.h"
+
+// To generate unicodetype_db.ll, fix the include above and compile it with
+//      clang -S -emit-llvm -O2 unicodetype_db.h -o unicodetype_db.ll
```
Comment (pearu) [inline rbc/unicodetype_db.h:5]: To avoid unnecessary changes to this file, I suggest:
```suggestion
#include "_unicodetype_db.h"

// To generate unicodetype_db.ll, fix the include above and compile it with
//      clang -S -emit-llvm -O2 -Ipath/to/numba/ unicodetype_db.h -o unicodetype_db.ll
```

File: rbc/irtools.py [hunk 21]
```diff
@@ -352,6 +394,14 @@ def add_metadata_flag(main_library, **kwargs):
     main_library.add_ir_module(module)
 
 
+def read_unicodetype_db():
+    unicode_file = os.path.join(os.path.dirname(__file__),
+                                'unicodetype_db.ll')
+    with open(unicode_file, 'r') as f:
+        s = ''.join(f.readlines())
+    return llvm.parse_assembly(s)
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/irtools.py:402]: Done!

File: rbc/irtools.py [hunk 22]
```diff
@@ -410,7 +466,7 @@ def compile_to_LLVM(functions_and_signatures,
         add_metadata_flag(main_library,
                           pass_column_arguments_by_value=0,
                           manage_memory_buffer=1)
-        main_library._optimize_final_module()
+        # main_library._optimize_final_module()
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/irtools.py]: Done!

File: rbc/tests/heavydb/test_nrt_set.py (hunk 19 above)
Comment (guilhermeleobas) [reply rbc/tests/heavydb/test_nrt_set.py:36]: It doesn't seem to work in Numba:
```python
self = <numba.core.typeinfer.TypeInferer object at 0x7ff2e5965f30>, raise_errors = True

    def propagate(self, raise_errors=True):
        newtoken = self.get_state_token()
        oldtoken = None
        # Since the number of types are finite, the typesets will eventually
        # stop growing.

        while newtoken != oldtoken:
            self.debug.propagate_started()
            oldtoken = newtoken
            # Errors can appear when the type set is incomplete; only
            # raise them when there is no progress anymore.
            errors = self.constraints.propagate(self)
            newtoken = self.get_state_token()
            self.debug.propagate_finished()
        if errors:
            if raise_errors:
                force_lit_args = [e for e in errors
                                  if isinstance(e, ForceLiteralArg)]
                if not force_lit_args:
>                   raise errors[0]
E                   numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E                   Invalid use of BoundFunction(set.union for set(int64)) with parameters (set(Literal[int](0)))
E
E                   During: resolving callee type: BoundFunction(set.union for set(int64))
E                   During: typing of call at /home/guilhermeleobas/git/rbc/rbc/tests/heavydb/test_nrt_set.py (84)
E
E
E                   File "rbc/tests/heavydb/test_nrt_set.py", line 84:
E                       def test_set(t, method):
E                           <source elided>
E                           elif method_str == 'union':
E                               return len(s.union({0})) == 6
E                               ^

../../miniconda3/envs/rbc/lib/python3.10/site-packages/numba/core/typeinfer.py:1086: TypingError
```

File: rbc/heavydb/text_encoding_none.py (hunk 11 above)
Comment (guilhermeleobas) [reply rbc/heavydb/text_encoding_none.py:198]: The code in the end is practically the same. For the sake of readability, let's keep `p1`.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @guilhermeleobas!
